File: src/server_llm/routes_chat.py
"""Chat endpoint for the LLM Gateway."""

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import Optional
from ..models import chat
from ..tools.memory_tools import soul_read
from ..config import PROMPTS_DIR
from .session_store import store
from .streaming import stream_chat

router = APIRouter(prefix="/api")


# --------------------------------------------------------------
class ChatRequest(BaseModel):
    """Incoming chat request."""
    message: str
    content: Optional[object] = None
    session_id: Optional[str] = None
    stream: Optional[bool] = True


# --------------------------------------------------------------
class ChatResponse(BaseModel):
    """Chat response."""
    reply: str
    session_id: Optional[str] = None


# --------------------------------------------------------------
def _load_system_prompt() -> str:
    """Load system prompt with soul context."""
    base = (
        PROMPTS_DIR / "SYSTEM.md"
    ).read_text(encoding="utf-8")
    soul = soul_read()
    return (
        f"{base}\n\n"
        f"# Core Identity (Soul)\n\n"
        f"{soul}"
    )


# --------------------------------------------------------------
def _build_messages(sid, system_prompt):
    """Build message list from session history."""
    history = store.get_messages(sid)
    return [
        {
            "role": "system",
            "content": system_prompt,
        },
        *history,
    ]


# --------------------------------------------------------------
@router.post("/chat")
async def chat_endpoint(req: ChatRequest):
    """Handle chat — streaming or non-streaming."""
    sid = req.session_id or "default"
    system_prompt = _load_system_prompt()

    # Determine user content (multipart or text)
    user_content = req.content or req.message

    # Store text-only version for session log
    store.add_message(sid, "user", req.message)

    # Build messages, replace last user msg
    # with rich content if files attached
    messages = _build_messages(
        sid, system_prompt
    )
    if req.content and isinstance(
        req.content, list
    ):
        # Replace last user message content
        # with multipart content array
        messages[-1] = {
            "role": "user",
            "content": req.content,
        }

    if req.stream:
        return _stream_response(sid, messages)
    return _sync_response(sid, messages)


# --------------------------------------------------------------
def _stream_response(sid, messages):
    """Return SSE streaming response."""
    collected = {"text": ""}

    async def generate():
        """Yield SSE events and save result."""
        async for line in stream_chat(messages):
            # Capture full_text event
            if '"full_text"' in line:
                import json
                data = line.replace(
                    "data: ", ""
                ).strip()
                obj = json.loads(data)
                collected["text"] = (
                    obj["full_text"]
                )
                store.add_message(
                    sid,
                    "assistant",
                    collected["text"],
                )
            yield line

    return StreamingResponse(
        generate(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "X-Accel-Buffering": "no",
        },
    )


# --------------------------------------------------------------
def _sync_response(sid, messages):
    """Return non-streaming JSON response."""
    resp = chat(messages)
    answer = (
        resp.choices[0].message.content or ""
    )
    store.add_message(sid, "assistant", answer)
    return ChatResponse(
        reply=answer,
        session_id=sid,
    )


# --------------------------------------------------------------
@router.get("/chat/history")
async def list_sessions():
    """List all session IDs."""
    return store.list_sessions()


# --------------------------------------------------------------
@router.get("/chat/{session_id}")
async def get_session(session_id: str):
    """Get messages for a session."""
    msgs = store.get_session(session_id)
    if msgs is None:
        raise HTTPException(
            status_code=404,
            detail="Session not found",
        )
    return msgs


# --------------------------------------------------------------
@router.delete("/chat/{session_id}")
async def delete_session(session_id: str):
    """Delete a session."""
    deleted = store.delete_session(session_id)
    if not deleted:
        raise HTTPException(
            status_code=404,
            detail="Session not found",
        )
    return {"status": "deleted"}

File: server_llm/src/routes_chat.rs
/// Chat endpoints — POST /api/chat (SSE), GET/DELETE /api/chat/{id}
use axum::{
    extract::{Path, State},
    routing::{delete, get, post},
    Json, Router,
};
use serde::Deserialize;
use serde_json::Value;
use tokio::sync::mpsc;
use tokio_stream::wrappers::ReceiverStream;
use tokio_stream::StreamExt as _;
use common::{config, tools::memory::soul_read};
use crate::session_store::AppState;
use crate::streaming::{stream_chat, non_stream_chat};

#[derive(Deserialize)]
pub struct ChatRequest {
    pub message: String,
    pub content: Option<Value>,
    pub session_id: Option<String>,
    pub stream: Option<bool>,
}

pub fn router(state: AppState) -> Router {
    Router::new()
        .route("/api/chat", post(chat_endpoint))
        .route("/api/chat/history", get(list_sessions))
        .route("/api/chat/{session_id}", get(get_session))
        .route("/api/chat/{session_id}", delete(delete_session))
        .with_state(state)
}

async fn load_system_prompt() -> String {
    let cfg = config::get();
    let base = std::fs::read_to_string(cfg.prompts_dir.join("SYSTEM.md")).unwrap_or_default();
    let soul = soul_read().unwrap_or_default();
    format!("{base}\n\n# Core Identity (Soul)\n\n{soul}")
}

async fn chat_endpoint(
    State(state): State<AppState>,
    Json(req): Json<ChatRequest>,
) -> axum::response::Response {
    let sid = req.session_id.clone().unwrap_or_else(|| "default".to_string());
    let system_prompt = load_system_prompt().await;

    state.store.add_message(&sid, "user", &req.message);

    let history = state.store.get_messages(&sid);
    let mut messages: Vec<Value> = vec![serde_json::json!({
        "role": "system",
        "content": system_prompt,
    })];
    for msg in &history {
        messages.push(serde_json::to_value(msg).unwrap_or_default());
    }
    if let Some(content) = &req.content {
        if let Some(last) = messages.last_mut() {
            if last["role"] == "user" {
                last["content"] = content.clone();
            }
        }
    }

    if req.stream.unwrap_or(true) {
        let (tx, rx) = mpsc::channel::<String>(256);

        tokio::spawn(async move {
            stream_chat(messages, tx.clone()).await;
        });

        // Wrap the channel into a byte stream, intercepting full_text to save to session
        let store2 = state.store.clone();
        let sid2 = sid.clone();
        let (byte_tx, byte_rx) = mpsc::channel::<Result<bytes::Bytes, std::convert::Infallible>>(256);

        tokio::spawn(async move {
            let mut recv_stream = ReceiverStream::new(rx);
            while let Some(line) = recv_stream.next().await {
                if line.contains("\"full_text\"") {
                    let data = line.trim_start_matches("data: ").trim();
                    if let Ok(obj) = serde_json::from_str::<Value>(data) {
                        if let Some(ft) = obj["full_text"].as_str() {
                            store2.add_message(&sid2, "assistant", ft);
                        }
                    }
                }
                let _ = byte_tx.send(Ok(bytes::Bytes::from(line))).await;
            }
        });

        let body = axum::body::Body::from_stream(ReceiverStream::new(byte_rx));
        axum::response::Response::builder()
            .header("Content-Type", "text/event-stream")
            .header("Cache-Control", "no-cache")
            .header("X-Accel-Buffering", "no")
            .body(body)
            .unwrap()
    } else {
        match non_stream_chat(messages).await {
            Ok(answer) => {
                state.store.add_message(&sid, "assistant", &answer);
                axum::response::Response::builder()
                    .header("Content-Type", "application/json")
                    .body(axum::body::Body::from(
                        serde_json::to_string(&serde_json::json!({
                            "reply": answer,
                            "session_id": sid,
                        }))
                        .unwrap(),
                    ))
                    .unwrap()
            }
            Err(e) => axum::response::Response::builder()
                .status(500)
                .body(axum::body::Body::from(e.to_string()))
                .unwrap(),
        }
    }
}

async fn list_sessions(State(state): State<AppState>) -> Json<Value> {
    Json(Value::Array(
        state.store.list_sessions()
            .into_iter()
            .map(Value::String)
            .collect(),
    ))
}

async fn get_session(
    State(state): State<AppState>,
    Path(session_id): Path<String>,
) -> axum::response::Response {
    match state.store.get_session(&session_id) {
        Some(msgs) => axum::response::Response::builder()
            .header("Content-Type", "application/json")
            .body(axum::body::Body::from(serde_json::to_string(&msgs).unwrap()))
            .unwrap(),
        None => axum::response::Response::builder()
            .status(404)
            .body(axum::body::Body::from(r#"{"detail":"Session not found"}"#))
            .unwrap(),
    }
}

async fn delete_session(
    State(state): State<AppState>,
    Path(session_id): Path<String>,
) -> axum::response::Response {
    if state.store.delete_session(&session_id) {
        axum::response::Response::builder()
            .header("Content-Type", "application/json")
            .body(axum::body::Body::from(r#"{"status":"deleted"}"#))
            .unwrap()
    } else {
        axum::response::Response::builder()
            .status(404)
            .body(axum::body::Body::from(r#"{"detail":"Session not found"}"#))
            .unwrap()
    }
}
